### DISCUSSION TITLE
Can I load an SQLite extension in GRDB?

### DISCUSSION THREAD

**Discussion by stsandro** [groue/GRDB.swift#1591] [Q&A]
I want to load an SQLite extension and then use it with raw SQL in GRDB. Is that possible, or am I ill-fated here?

**Comment by groue** [comment]
Hello @stsandro, my Google search tells that the SQLite library shipped on Apple operating systems is not able to load extensions.

You can link GRDB with a [custom SQLite build](https://github.com/groue/GRDB.swift/blob/master/Documentation/CustomSQLiteBuilds.md) that is able to load extensions. This is uncharted territory, so please report the results of your investigations if you have time.

**Reply by michaelmedellin** [reply]
@groue @stsandro  it is possible with a custom SQLite build, however the SQLiteLib project used to build SQLite actually mirrors the configuration of the onboard iOS SQLite that disables extension loading via `-DSQLITE_OMIT_LOAD_EXTENSION` flag passed during compilation [line reference here](https://github.com/swiftlyfalling/SQLiteLib/blob/31e6aa66188e59616f062df77329d6ee9ee45929/SQLiteLib.xcconfig#L78).

The SQLiteLib project doesn't allow for removing of that flag as user provided configuration, so I've forked it to remove the flag and loading extensions works. [I added an issue in that project asking about remove flags as user customization.](https://github.com/swiftlyfalling/SQLiteLib/issues/61#issue-2420899368)

**Reply by groue** [reply]
Thank you @michaelmedellin! Let's wait a little bit, until @swiftlyfalling can see your issue.

**Reply by stsandro** [reply]
@groue @michaelmedellin Thank you both for the information provided. I'll have a closer look at everything when I get back from vacation.

**Reply by spprichard** [reply]
Hey all, 

I'm curious if there has been any movement on this issue? 

I think if there is a way to load an extension with GRDB, some really exciting use cases could be unlocked. For example using the [sqlite-vec](https://github.com/asg017/sqlite-vec) extension for on-device vector databases (assuming you could address the SQLite issues highlighted above)

**Reply by groue** [reply]
Hello @spprichard, I did not take the time to remove `SQLITE_OMIT_LOAD_EXTENSION` from http://github.com/swiftlyfalling/SQLiteLib.

Is it possible that you would submit a pull request to it, after having checked that it indeed allows you to complete the picture (build an app that uses a custom SQLite build, loads an extension, and uses GRDB)?

Extra bonus: a second pull request on the [Custom SQLite Builds](https://github.com/groue/GRDB.swift/blob/master/Documentation/CustomSQLiteBuilds.md) GRDB guide that explains how to load an extension. A down-to-earth step-by-step guide, illustrated with the extension of your choice, will provide the guidance people need. A very explicit tone is better: the less people have to perform choices or guesses, the better. Please do not assume any familiarity with the topic! For many people who need an extension, it is the very first time, and they are totally uneducated (I am myself in this situation 😉) I like GRDB documentation to _also_ teach about SQLite.

**Reply by spprichard** [reply]
I'll give it a shot! 

I'll be honest tho, I don't fully understand how & why GRDB has this dependency on SQLiteLib in the way in which it does (as an embedded xcode proj?). Is there any documentation discussing the rational? I'm sure there are completely valid reasons, I am just uneducated and looking to understand. I am also new-ish to Swift development so I may not know how things have been done historically, and I am looking to learn. 😄

**Reply by groue** [reply]
> I'll give it a shot! 

That would be great!

> I'll be honest tho, I don't fully understand how & why GRDB has this dependency on SQLiteLib in the way in which it does (as an embedded xcode proj?). Is there any documentation discussing the rational? I'm sure there are completely valid reasons, I am just uneducated and looking to understand. I am also new-ish to Swift development so I may not know how things have been done historically, and I am looking to learn. 😄

The original contributor answers some of your questions in the pull request #62.

From my own perspective, if was a boon. Suddenly it was possible to build and customize SQLite with a ready-made recipe (not _simple_, I'll grant you that, but _regular_ - another kind of simplicity I've learned to appreciate). Also, at the time, the Swift package manager did not exist. Building with Xcode wasn't as surprising as it might be today.

To be honest, SPM is far from being able to do the beginning of the start of the quarter of what Xcode can do. SQLite and SQLCipher are not available through SPM, and even if SPM could _build_ them (which it can not: https://github.com/sqlcipher/sqlcipher/issues/371), _customization_ is yet another (future) story. On the customization side, [Package Traits](https://forums.swift.org/t/se-0450-package-traits/75598), currently on review, might help - but I wouldn't hold my breath because SQLite does not look like it is present in the consciousness of the designers. You might want to chime in those forums.

Anyway. SPM being out of the game due to its young age, betting on Xcode sounds like a rational option to me. And SQLiteLib has a tremendous power of customization (for the GRDB user), associated with very simple maintenance (for me). I barely had to change anything since 2016, but download, unzip, and commit the latest SQLite sources from time to time. It's... _perfect_. @swiftlyfalling has been an angel.

(And in case it's not very clear: I'm not fond of _building_. I do not really appreciate to tickle knobs, dig ancient Apple tech notes that are the only source for necessary information, search Google for obscure errors, summon ChatGPT and be disappointed. Hence my focus on regular recipes - developers who lack taste for building love them.)

**Reply by spprichard** [reply]
Thank you for the detailed response! There's a lot there to parse. 

Forgive me if I am speaking out of me depths here. I was looking over the way the [swift-toolchain-sqlite](https://github.com/swiftlang/swift-toolchain-sqlite/blob/main/Package.swift) constructs their package, and was shocked how simple it is. 

From what I can tell, all the C code is in its own target, which has its own process for updating as new versions of SQLite become available. There's no shims, bridging, or module maps. I could be wrong here, but I think constructing the SQLite dependent package this was make use of the C interop within Swift. 

Maybe this is over simplistic & for the reasons you state above this won't work, but could GRDB make a similar type package (or refactor the SQLiteLib dependency) to follow this? In my head, this would simplify  the custom SQLite build story for GRDB.

**Reply by groue** [reply]
> Forgive me if I am speaking out of me depths here. I was looking over the way the [swift-toolchain-sqlite](https://github.com/swiftlang/swift-toolchain-sqlite/blob/main/Package.swift) constructs their package, and was shocked how simple it is. 

You're all forgiven.

When I look at the [content of the CSQLite target](https://github.com/swiftlang/swift-toolchain-sqlite/tree/main/Sources/CSQLite), I see a single sqlite.c file. Because it is the only file, I know it is an [SQLite amalgamation](https://sqlite.org/amalgamation.html). It is the result of the processing of the (many) raw SQLite C files, according to a chosen and fixed set of [compile-time options](https://sqlite.org/compile.html). IIUC, those options also change the output of the Makefile that build the sqlite.c amalgamation (it's not only a matter of concatenating a bunch of files together). But even if I'm wrong I this last sentence, this does not change the net result:

The sqlite.c file _is not_ built by the swiftlang/swift-toolchain-sqlite package. It is pre-built, and only configurable by the authors of the package. It is not configurable by the package users.

SQLiteLib, on the other hand, allows user configuration.

> Maybe this is over simplistic & for the reasons you state above this won't work, but could GRDB make a similar type package (or refactor the SQLiteLib dependency) to follow this? In my head, this would simplify  the custom SQLite build story for GRDB.

I could ship GRDB with a custom build of SQLite, yes, by following the same technique. It would be fixed on an arbitrary set of compile-time options. I would chose a set that makes consensus, of course, but it would not be configurable by GRDB users. People could not remove code that they don't use in order to reduce the size of the build, or add the features they need.

The README of the Swift SQLite package says:

> This is intended as a common infrastructure package for components of the Swift toolchain which depend on SQLite.

Read: it is built with what we need for the Swift toolchain, nothing more, nothing less, and do not ask for more features because:

> It is not intended as a general-purpose SQLite package.

If GRDB followed the same route, we'd have the same result: a rigid version of SQLite.

SQLiteLib, if more involved, is more versatile. It addresses the needs of demanding developers, and this fits pretty well the spirit of GRDB.

**Reply by spprichard** [reply]
That makes total sense. Thank you!

**Reply by spprichard** [reply]
Wanted to follow up on this. 

I am a bit stuck/lost. I _think_ I have followed the directions within SQLiteLib & GRDB for how to embed/update SQLiteLib. However, I am getting this build error for the GRDBCustom scheme. 

<img width="998" alt="image" src="https://github.com/user-attachments/assets/76fcaf5a-7710-49d9-927f-d00fbace3a99">

What I did to get into this state
- Open GRDB workspace
- Within the GRDBCustom project, delete the SQLiteLib project (GRDBCustomSQLite/SQLite.xcproj)
- Drag and drop the version of SQLiteLib xcode project with `OMIT_LOAD_EXTENSION` removed from `SQLiteLib.xcconfig`. Here's the [draft PR](https://github.com/swiftlyfalling/SQLiteLib/pull/64) for that change to SQLiteLib. 
- Attempt to build the `GRDBCustom` scheme

My initial thought was to just add the file its complaining about, but that doesn't feel right to me. Because this wasn't failing prior to my change, so its probably something I did (or haven't done) within my changes thats causing the error

**Reply by groue** [reply]
Thanks for the detailed report. I suggest we move our discussion to https://github.com/groue/GRDB.swift/pull/1680 from now on.
